working/1113 - null-asserting compound assignment/proposal.md

```diff
@@ -0,0 +1,160 @@
+# Dart Null-Asserting Composite Assignment
+
+Author: lrn@google.com<br>Version: 0.1
+
+## Background and Motivation
+
+With the current Null Safety behavior, a nullable variable like `int? x;` cannot be incremented as easily
+as a non-nullable variable, even when you know that it currently contains a non-null value.
+We have `++x` and `x += 1` as easy composite operations on non-nullable types,
+but for the nullable type, when you know it's not null, the only option is `x = x! + 1` 
+because we need to insert the `!` in the middle of the expression that `++x` or `x += 1` expands to.
+
+This document proposes a new syntax, `x! += 1`, `++x!` or `x!++`, 
+which is a short-hand for an expanded version like the one above, 
+but where the *read* of the value is guarded with a non-`null` asserting check.
+
+### Composite operator specification 
+
+Currently we have a multiplicity of rules for various assignable expressions/left-hand sides (LHSs):
+
+```dart
+id
+e1.id
+e1?.id
+e1[e2]
+e1?[e2]
+super.id
+super[e2]
+ClassName.id  
+```
```

Comment:
These forms are of course assumed to be derivable in the grammar, so `e1` is a `(<constructorInvocation>|<primary>) <selector>*`, and `ClassName` could be a subset of that (just `<identifier>|<qualified>`), with the constraint that it is a type literal (so it dominates `e1.id` when applicable, and should perhaps occur earlier in the list).

Reply:
ACK